File: 05/gpt-5.1/python/solution.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class Interval:
    start: int
    end: int


with open("input.txt", encoding="utf-8") as f:
    lines = [line.strip() for line in f.read().splitlines() if line.strip()]

ranges: list[Interval] = []
available_ids: list[int] = []

index = 0
while index < len(lines) and "-" in lines[index]:
    line = lines[index]
    left, right = (part.strip() for part in line.split("-", 1))
    start = int(left)
    end = int(right)
    if end < start:
        start, end = end, start
    ranges.append(Interval(start, end))
    index += 1

while index < len(lines):
    line = lines[index]
    if line:
        available_ids.append(int(line))
    index += 1

ranges.sort(key=lambda r: (r.start, r.end))

merged: list[Interval] = []
for r in ranges:
    if not merged:
        merged.append(Interval(r.start, r.end))
        continue
    last = merged[-1]
    if r.start <= last.end + 1:
        if r.end > last.end:
            last.end = r.end
    else:
        merged.append(Interval(r.start, r.end))

part1 = 0
for value in available_ids:
    for rng in merged:
        if value < rng.start:
            break
        if value <= rng.end:
            part1 += 1
            break

print(f"Day 05 Part 1: {part1}")

part2 = 0
for rng in merged:
    part2 += rng.end - rng.start + 1

print(f"Day 05 Part 2: {part2}")

File: 05/gpt-5.1/typescript/solution.ts
import * as fs from 'fs';

const raw = fs.readFileSync('input.txt', 'utf-8');
const lines = raw.split(/\r?\n/).filter(l => l.length > 0);

type Range = { start: number; end: number };

const ranges: Range[] = [];
const available: number[] = [];

let index = 0;
while (index < lines.length && lines[index].includes('-')) {
	const line = lines[index].trim();
	if (line.length > 0) {
		const [s, e] = line.split('-', 2).map(part => part.trim());
		const start = Number(s);
		const end = Number(e);
		if (!Number.isNaN(start) && !Number.isNaN(end)) {
			const rStart = Math.min(start, end);
			const rEnd = Math.max(start, end);
			ranges.push({ start: rStart, end: rEnd });
		}
	}
	index++;
}

for (; index < lines.length; index++) {
	const line = lines[index].trim();
	if (line.length === 0) continue;
	const id = Number(line);
	if (!Number.isNaN(id)) {
		available.push(id);
	}
}

ranges.sort((a, b) => (a.start === b.start ? a.end - b.end : a.start - b.start));

const merged: Range[] = [];
for (const r of ranges) {
	if (merged.length === 0) {
		merged.push({ ...r });
		continue;
	}
	const last = merged[merged.length - 1];
	if (r.start <= last.end + 1) {
		if (r.end > last.end) {
			last.end = r.end;
		}
	} else {
		merged.push({ ...r });
	}
}

let part1 = 0;
for (const id of available) {
	for (const r of merged) {
		if (id < r.start) break;
		if (id <= r.end) {
			part1++;
			break;
		}
	}
}
console.log(`Day 05 Part 1: ${part1}`);

let part2 = 0;
for (const r of merged) {
	part2 += r.end - r.start + 1;
}
console.log(`Day 05 Part 2: ${part2}`);